### ISSUE TITLE
Undefined Reference Error

### ISSUE THREAD

**Issue by CyberDrudge** [STEllAR-GROUP/hpx#3592]
On Kali-Linux
Setting up a new HPX project:
```
cmake -DCMAKE_PREFIX_PATH=/Doc/hpx/ ..
make all
```
Gives Error:
```
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `main':
main.cpp:(.text+0x44): undefined reference to `hpx::cout'
/usr/bin/ld: main.cpp:(.text+0x50): undefined reference to `hpx::iostreams::flush'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `void hpx::lcos::packaged_action<hpx::iostreams::server::output_stream::write_sync_action, void, false>::do_apply<unsigned int, unsigned long, hpx::iostreams::detail::buffer&>(hpx::naming::address&&, hpx::naming::id_type const&, hpx::threads::thread_priority, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&)':
main.cpp:(.text._ZN3hpx4lcos15packaged_actionINS_9iostreams6server13output_stream17write_sync_actionEvLb0EE8do_applyIJjmRNS2_6detail6bufferEEEEvONS_6naming7addressERKNSB_7id_typeENS_7threads15thread_priorityEDpOT_[_ZN3hpx4lcos15packaged_actionINS_9iostreams6server13output_stream17write_sync_actionEvLb0EE8do_applyIJjmRNS2_6detail6bufferEEEEvONS_6naming7addressERKNSB_7id_typeENS_7threads15thread_priorityEDpOT_]+0xd6): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `bool hpx::apply_p_cb<hpx::iostreams::server::output_stream::write_sync_action, hpx::actions::typed_continuation<void, hpx::util::unused_type>, hpx::util::detail::bound_front<void (*)(boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> >, boost::system::error_code const&, hpx::parcelset::parcel const&), boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> > >, unsigned int, unsigned long, hpx::iostreams::detail::buffer&>(hpx::actions::typed_continuation<void, hpx::util::unused_type>&&, hpx::naming::address&&, hpx::naming::id_type const&, hpx::threads::thread_priority, hpx::util::detail::bound_front<void (*)(boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> >, boost::system::error_code const&, hpx::parcelset::parcel const&), boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> > >&&, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&)':
main.cpp:(.text._ZN3hpx10apply_p_cbINS_9iostreams6server13output_stream17write_sync_actionENS_7actions18typed_continuationIvNS_4util11unused_typeEEENS7_6detail11bound_frontIPFvN5boost13intrusive_ptrINS_4lcos6detail12promise_dataIvEEEERKNSC_6system10error_codeERKNS_9parcelset6parcelEEJSI_EEEJjmRNS1_6detail6bufferEEEEbOT0_ONS_6naming7addressERKNSZ_7id_typeENS_7threads15thread_priorityEOT1_DpOT2_[_ZN3hpx10apply_p_cbINS_9iostreams6server13output_stream17write_sync_actionENS_7actions18typed_continuationIvNS_4util11unused_typeEEENS7_6detail11bound_frontIPFvN5boost13intrusive_ptrINS_4lcos6detail12promise_dataIvEEEERKNSC_6system10error_codeERKNS_9parcelset6parcelEEJSI_EEEJjmRNS1_6detail6bufferEEEEbOT0_ONS_6naming7addressERKNSZ_7id_typeENS_7threads15thread_priorityEOT1_DpOT2_]+0xb0): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::basic_action_impl<void (hpx::iostreams::server::output_stream::*)(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&), void (hpx::iostreams::server::output_stream::*)(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&), &hpx::iostreams::server::output_stream::write_sync, hpx::iostreams::server::output_stream::write_sync_action>::get_action_name[abi:cxx11](unsigned long)':
main.cpp:(.text._ZN3hpx7actions17basic_action_implIMNS_9iostreams6server13output_streamEFvjmRKNS2_6detail6bufferEESA_XadL_ZNS4_10write_syncEjmS8_EENS4_17write_sync_actionEE15get_action_nameB5cxx11Em[_ZN3hpx7actions17basic_action_implIMNS_9iostreams6server13output_streamEFvjmRKNS2_6detail6bufferEESA_XadL_ZNS4_10write_syncEjmS8_EENS4_17write_sync_actionEE15get_action_nameB5cxx11Em]+0x5a): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `std::enable_if<hpx::traits::is_continuation<hpx::actions::typed_continuation<void, hpx::util::unused_type>, void>::value, bool>::type hpx::detail::apply_cb_impl<hpx::iostreams::server::output_stream::write_sync_action, hpx::actions::typed_continuation<void, hpx::util::unused_type>, hpx::util::detail::bound_front<void (*)(boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> >, boost::system::error_code const&, hpx::parcelset::parcel const&), boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> > >, unsigned int, unsigned long, hpx::iostreams::detail::buffer&>(hpx::actions::typed_continuation<void, hpx::util::unused_type>&&, hpx::naming::id_type const&, hpx::threads::thread_priority, hpx::util::detail::bound_front<void (*)(boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> >, boost::system::error_code const&, hpx::parcelset::parcel const&), boost::intrusive_ptr<hpx::lcos::detail::promise_data<void> > >&&, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&)':
main.cpp:(.text._ZN3hpx6detail13apply_cb_implINS_9iostreams6server13output_stream17write_sync_actionENS_7actions18typed_continuationIvNS_4util11unused_typeEEENS8_6detail11bound_frontIPFvN5boost13intrusive_ptrINS_4lcos6detail12promise_dataIvEEEERKNSD_6system10error_codeERKNS_9parcelset6parcelEEJSJ_EEEJjmRNS2_6detail6bufferEEEENSt9enable_ifIXsrNS_6traits15is_continuationIT0_vEE5valueEbE4typeEOS11_RKNS_6naming7id_typeENS_7threads15thread_priorityEOT1_DpOT2_[_ZN3hpx6detail13apply_cb_implINS_9iostreams6server13output_stream17write_sync_actionENS_7actions18typed_continuationIvNS_4util11unused_typeEEENS8_6detail11bound_frontIPFvN5boost13intrusive_ptrINS_4lcos6detail12promise_dataIvEEEERKNSD_6system10error_codeERKNS_9parcelset6parcelEEJSJ_EEEJjmRNS2_6detail6bufferEEEENSt9enable_ifIXsrNS_6traits15is_continuationIT0_vEE5valueEbE4typeEOS11_RKNS_6naming7id_typeENS_7threads15thread_priorityEOT1_DpOT2_]+0xb0): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `int hpx::components::get_component_type<hpx::iostreams::server::output_stream>()':
main.cpp:(.text._ZN3hpx10components18get_component_typeINS_9iostreams6server13output_streamEEEiv[_ZN3hpx10components18get_component_typeINS_9iostreams6server13output_streamEEEiv]+0x5): undefined reference to `hpx::traits::component_type_database<hpx::iostreams::server::output_stream, void>::get()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `std::enable_if<!hpx::traits::is_future<void>::value, void>::type hpx::actions::basic_action_impl<void (hpx::iostreams::server::output_stream::*)(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&), void (hpx::iostreams::server::output_stream::*)(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&), &hpx::iostreams::server::output_stream::write_sync, hpx::iostreams::server::output_stream::write_sync_action>::invoke_helper<void, unsigned int, unsigned long, hpx::iostreams::detail::buffer&>(hpx::actions::detail::tag<void>, unsigned long, int, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&)':
main.cpp:(.text._ZN3hpx7actions17basic_action_implIMNS_9iostreams6server13output_streamEFvjmRKNS2_6detail6bufferEESA_XadL_ZNS4_10write_syncEjmS8_EENS4_17write_sync_actionEE13invoke_helperIvJjmRS6_EEENSt9enable_ifIXntsrNS_6traits9is_futureIT_EE5valueESI_E4typeENS0_6detail3tagISI_EEmiDpOT0_[_ZN3hpx7actions17basic_action_implIMNS_9iostreams6server13output_streamEFvjmRKNS2_6detail6bufferEESA_XadL_ZNS4_10write_syncEjmS8_EENS4_17write_sync_actionEE13invoke_helperIvJjmRS6_EEENSt9enable_ifIXntsrNS_6traits9is_futureIT_EE5valueESI_E4typeENS0_6detail3tagISI_EEmiDpOT0_]+0x6c): undefined reference to `hpx::iostreams::server::output_stream::write_sync(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&)'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::transfer_continuation_action<hpx::iostreams::server::output_stream::write_sync_action>::transfer_continuation_action<hpx::threads::thread_priority&, unsigned int, unsigned long, hpx::iostreams::detail::buffer&>(hpx::actions::typed_continuation<void, hpx::util::unused_type>&&, hpx::threads::thread_priority&, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&)':
main.cpp:(.text._ZN3hpx7actions28transfer_continuation_actionINS_9iostreams6server13output_stream17write_sync_actionEEC2IJRNS_7threads15thread_priorityEjmRNS2_6detail6bufferEEEEONS0_18typed_continuationIvNS_4util11unused_typeEEEDpOT_[_ZN3hpx7actions28transfer_continuation_actionINS_9iostreams6server13output_stream17write_sync_actionEEC5IJRNS_7threads15thread_priorityEjmRNS2_6detail6bufferEEEEONS0_18typed_continuationIvNS_4util11unused_typeEEEDpOT_]+0x7c): undefined reference to `vtable for hpx::actions::transfer_continuation_action<hpx::iostreams::server::output_stream::write_sync_action>'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::transfer_continuation_action<hpx::iostreams::server::output_stream::write_sync_action>::transfer_continuation_action<hpx::threads::thread_priority, unsigned int, unsigned long, hpx::iostreams::detail::buffer>(hpx::actions::typed_continuation<void, hpx::util::unused_type>&&, hpx::threads::thread_priority&&, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&&)':
main.cpp:(.text._ZN3hpx7actions28transfer_continuation_actionINS_9iostreams6server13output_stream17write_sync_actionEEC2IJNS_7threads15thread_priorityEjmNS2_6detail6bufferEEEEONS0_18typed_continuationIvNS_4util11unused_typeEEEDpOT_[_ZN3hpx7actions28transfer_continuation_actionINS_9iostreams6server13output_stream17write_sync_actionEEC5IJNS_7threads15thread_priorityEjmNS2_6detail6bufferEEEEONS0_18typed_continuationIvNS_4util11unused_typeEEEDpOT_]+0x7c): undefined reference to `vtable for hpx::actions::transfer_continuation_action<hpx::iostreams::server::output_stream::write_sync_action>'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::detail::register_action<hpx::iostreams::server::output_stream::write_sync_action>::register_action()':
main.cpp:(.text._ZN3hpx7actions6detail15register_actionINS_9iostreams6server13output_stream17write_sync_actionEEC2Ev[_ZN3hpx7actions6detail15register_actionINS_9iostreams6server13output_stream17write_sync_actionEEC5Ev]+0x22): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::transfer_continuation_action<hpx::iostreams::server::output_stream::write_sync_action>::transfer_continuation_action()':
main.cpp:(.text._ZN3hpx7actions28transfer_continuation_actionINS_9iostreams6server13output_stream17write_sync_actionEEC2Ev[_ZN3hpx7actions28transfer_continuation_actionINS_9iostreams6server13output_stream17write_sync_actionEEC5Ev]+0x1c): undefined reference to `vtable for hpx::actions::transfer_continuation_action<hpx::iostreams::server::output_stream::write_sync_action>'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::transfer_action<hpx::iostreams::server::output_stream::write_sync_action>::transfer_action()':
main.cpp:(.text._ZN3hpx7actions15transfer_actionINS_9iostreams6server13output_stream17write_sync_actionEEC2Ev[_ZN3hpx7actions15transfer_actionINS_9iostreams6server13output_stream17write_sync_actionEEC5Ev]+0x1b): undefined reference to `vtable for hpx::actions::transfer_action<hpx::iostreams::server::output_stream::write_sync_action>'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `std::enable_if<!hpx::traits::is_future<void>::value, void>::type hpx::actions::basic_action_impl<void (hpx::iostreams::server::output_stream::*)(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&), void (hpx::iostreams::server::output_stream::*)(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&), &hpx::iostreams::server::output_stream::write_sync, hpx::iostreams::server::output_stream::write_sync_action>::invoke_helper<void, unsigned int, unsigned long, hpx::iostreams::detail::buffer>(hpx::actions::detail::tag<void>, unsigned long, int, unsigned int&&, unsigned long&&, hpx::iostreams::detail::buffer&&)':
main.cpp:(.text._ZN3hpx7actions17basic_action_implIMNS_9iostreams6server13output_streamEFvjmRKNS2_6detail6bufferEESA_XadL_ZNS4_10write_syncEjmS8_EENS4_17write_sync_actionEE13invoke_helperIvJjmS6_EEENSt9enable_ifIXntsrNS_6traits9is_futureIT_EE5valueESH_E4typeENS0_6detail3tagISH_EEmiDpOT0_[_ZN3hpx7actions17basic_action_implIMNS_9iostreams6server13output_streamEFvjmRKNS2_6detail6bufferEESA_XadL_ZNS4_10write_syncEjmS8_EENS4_17write_sync_actionEE13invoke_helperIvJjmS6_EEENSt9enable_ifIXntsrNS_6traits9is_futureIT_EE5valueESH_E4typeENS0_6detail3tagISH_EEmiDpOT0_]+0x6c): undefined reference to `hpx::iostreams::server::output_stream::write_sync(unsigned int, unsigned long, hpx::iostreams::detail::buffer const&)'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `hpx::actions::transfer_base_action<hpx::iostreams::server::output_stream::write_sync_action>::get_action_name() const':
main.cpp:(.text._ZNK3hpx7actions20transfer_base_actionINS_9iostreams6server13output_stream17write_sync_actionEE15get_action_nameEv[_ZNK3hpx7actions20transfer_base_actionINS_9iostreams6server13output_stream17write_sync_actionEE15get_action_nameEv]+0xd): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
/usr/bin/ld: CMakeFiles/my_hpx_program.dir/main.cpp.o: in function `unsigned int hpx::actions::detail::get_action_id<hpx::iostreams::server::output_stream::write_sync_action>()':
main.cpp:(.text._ZN3hpx7actions6detail13get_action_idINS_9iostreams6server13output_stream17write_sync_actionEEEjv[_ZN3hpx7actions6detail13get_action_idINS_9iostreams6server13output_stream17write_sync_actionEEEjv]+0x33): undefined reference to `char const* hpx::actions::detail::get_action_name<hpx::iostreams::server::output_stream::write_sync_action>()'
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/my_hpx_program.dir/build.make:100: my_hpx_program] Error 1
make[1]: *** [CMakeFiles/Makefile2:73: CMakeFiles/my_hpx_program.dir/all] Error 2
make: *** [Makefile:130: all] Error 2
```
CMakeLists.cpp :
```
cmake_minimum_required(VERSION 3.3.2)
project(my_hpx_project CXX)
find_package(HPX REQUIRED)
add_hpx_executable(my_hpx_program SOURCES main.cpp)
```
Main.cpp:
```
#include <hpx/hpx_main.hpp>
#include <hpx/include/iostreams.hpp>

int main()
{
    hpx::cout << "Hello World!\n" << hpx::flush;
    return 0;
}
```

**Comment by NK-Nikunj** [comment]
@CyberDrudge you need to add iostream component as a dependency.

Change the **add_hpx_executable** to:
`add_hpx_executable(my_hpx_program SOURCES main.cpp COMPONENT_DEPENDENCIES iostreams)`

**Comment by msimberg** [comment]
@CyberDrudge, I'm guessing you copied this from the quick start guide? This has come up before, and I'm getting the feeling the instructions are not as clear as they could be. Would it have been easier for you to start if we would just have included the `iostreams` dependency in the first `CMakeLists.txt` example?

On a more meta-level, this is a problem that would (in a sense) be solved by #3546. It shouldn't be possible to include `iostreams.hpp` without linking to the library at the same time.

**Comment by hkaiser** [comment]
I think we can close this. Please reopen if the problem persists.
